dtm0/fop.c

```diff
@@ -244,154 +206,177 @@ static size_t dtm0_fom_locality(const struct m0_fom *fom)
 	return locality++;
 }
 
-static int m0_dtm0_send_msg(struct m0_fom                *fom,
-			    enum m0_dtm0s_msg             msg_type,
-			    const struct m0_fid          *tgt,
-			    const struct m0_dtm0_tx_desc *txd)
+static int dtm0_req_post(struct m0_dtm0_service       *dtms,
+			 enum m0_dtm0s_msg             msg_type,
+			 const struct m0_dtm0_tx_desc *txd,
+			 const struct m0_fid          *tgt,
+			 struct m0_fom                *fom)
 {
-	struct m0_fop          *fop;
-	struct m0_rpc_session  *session;
-	struct m0_rpc_item     *item;
-	struct dtm0_req_fop    *req;
-	uint64_t                phase_sm_id;
-	uint64_t                rpc_sm_id;
-	int                     rc;
-	struct m0_dtm0_service *dtms = m0_dtm0_service_find(
-		fom->fo_service->rs_reqh);
-
-	M0_ENTRY("reqh=%p", dtms->dos_generic.rs_reqh);
-
-	session = m0_dtm0_service_process_session_get(&dtms->dos_generic, tgt);
-	if (session == NULL)
-		return M0_ERR(-ENOENT);
-
-	fop = m0_fop_alloc_at(session, &dtm0_req_fop_fopt);
-	if (fop == NULL)
-		return M0_ERR(-ENOMEM);
-
-	item              = &fop->f_item;
-	item->ri_ops      = &dtm0_req_fop_rpc_item_ops;
-	item->ri_session  = session;
-	item->ri_prio     = M0_RPC_ITEM_PRIO_MID;
-	item->ri_deadline = M0_TIME_IMMEDIATELY;
-
-	req               = m0_fop_data(fop);
-	req->dtr_msg      = msg_type;
-	rc = m0_dtm0_tx_desc_copy(txd, &req->dtr_txr) ?: m0_rpc_post(item);
-	if (rc == 0) {
-		phase_sm_id = m0_sm_id_get(&fom->fo_sm_phase);
-		rpc_sm_id   = m0_sm_id_get(&item->ri_sm);
-		M0_ADDB2_ADD(M0_AVI_FOM_TO_TX, phase_sm_id, rpc_sm_id);
-
-		m0_fop_put_lock(fop);
-
-		M0_LOG(M0_DEBUG, "Sent %d msg " FID_F " -> " FID_F, msg_type,
-		       FID_P(&dtms->dos_generic.rs_service_fid), FID_P(tgt));
-	}
-	return M0_RC(rc);
+	return m0_dtm0_req_post(dtms,
+				&((struct dtm0_req_fop) {
+				  .dtr_msg = msg_type,
+				  .dtr_txr = *txd
+				  }),
+				tgt, fom, false);
```

Comment:
Yes, that makes sense to drop `dtm0_req_post` (see https://github.com/Seagate/cortx-motr/pull/802/commits/c75464d7d2ce0553ee63b8fa00745944e6dee137#diff-7d29e2d404e16606678795ff91d5da8bc5dbc20695239029de046a0ac8bb622d). However, I'd rather keep `m0_dtm0_req_post` as it is. If `m0_dtm0_req_post` required msg+txd (as dtm0_req_post) then it means this function would need to be rewritten whenever dtm0_req_fop has been changed (for example, if we added .payload field there). So that, I'd let `m0_dtm0_req_post` operate on the whole dtm0_req_fop rather than on its pieces (msg, txr and so on).